python/ray/tune/experiment/trial.py

```diff
@@ -1104,7 +1104,7 @@ def on_checkpoint(self, checkpoint: Union[_TrackedCheckpoint, _TrainingResult]):
             # Increment the checkpoint index to keep the checkpoint index in sync.
             # This index will get restored when the trial is restored and will
             # be passed to the Trainable as the starting checkpoint index.
-            self.storage.current_checkpoint_index += 1
+            self.storage._increase_checkpoint_index(checkpoint_result.metrics)
```

Comment:
```suggestion
            self.storage.current_checkpoint_index = checkpoint.result.metrics[TRAINING_ITERATION]
```

Reply:
Doesn't this mean that the checkpoint manager in `Trial` will get out of sync with class trainables?

Reply:
Is this the only location where we access the trial checkpoint ID?

```
            result[CHECKPOINT_DIR_NAME] = (
                trial.storage.checkpoint_dir_name if trial_will_checkpoint else None
            )
```

Reply:
Ah yeah, that's right.

That is the only place it gets explicitly accessed, but the trial checkpoint id also needs to stay in sync for restore -- it'll get propagated to the new trainable that gets scheduled:

https://github.com/ray-project/ray/blob/5369febefcdffc983f69e12d578d76557edad46d/python/ray/tune/experiment/trial.py#L277

Reply:
OK - this means we should keep those in sync and shouldn't diverge between class and function trainable, right?

In that case, should we go with the workaround proposed in this PR?

Alternatively, we could communicate the (updated) checkpoint ID in the training result.